src/internal/Subscription.ts

```diff
@@ -166,6 +166,8 @@ export class Subscription implements SubscriptionLike {
           const tmp = subscription;
           subscription = new Subscription();
           subscription._subscriptions = [tmp];
+        } else if (this._subscriptions && this._subscriptions.some(s => s === subscription)) {
+          return subscription;
         }
```

Comment:
I don't see why this is necessary. How does this relate to the change you made in `Observable.ts`?

Reply:
This change prevents three unit tests from failing. Otherwise the change in `Observable.ts` would add `Subscribers` a second time to the `_subscriptions` array, because most built-in operators have their own `Subscriber` implementation which already add themself to the array: https://github.com/ReactiveX/rxjs/blob/ef557bb1c3e8a56540fb117b4e1ce64a28276a40/src/internal/Subscriber.ts#L76

Here are the failing unit tests:
```
3 failing

  1) observeOn operator
       should clean up subscriptions created by async scheduling (prevent memory leaks #2244):

      Uncaught AssertionError: expected 3 to equal 2
      + expected - actual

      -3
      +2
      
      at SafeSubscriber._next (spec/operators/observeOn-spec.ts:108:57)
      at SafeSubscriber.__tryOrUnsub (src/internal/Subscriber.ts:267:10)
      at SafeSubscriber.next (src/internal/Subscriber.ts:209:14)
      at Subscriber._next (src/internal/Subscriber.ts:139:22)
      at Subscriber.next (src/internal/Subscriber.ts:99:12)
      at Notification.observe (src/internal/Notification.ts:36:42)
      at AsapAction.ObserveOnSubscriber.dispatch (src/internal/operators/observeOn.ts:81:18)
      at AsapAction.AsyncAction._execute (src/internal/scheduler/AsyncAction.ts:121:12)
      at AsapAction.AsyncAction.execute (src/internal/scheduler/AsyncAction.ts:96:24)
      at AsapScheduler.flush (src/internal/scheduler/AsapScheduler.ts:17:26)
      at runIfPresent (src/internal/util/Immediate.ts:8:5)
      at src/internal/util/Immediate.ts:16:34
      at <anonymous>

  2) switchAll
       should not leak when child completes before each switch (prevent memory leaks #2355):

      AssertionError: expected 2 to equal 1
      + expected - actual

      -2
      +1
      
      at spec/operators/switch-spec.ts:239:10
      at Context.modified (spec/helpers/testScheduler-ui.ts:206:13)

  3) switchAll
       should not leak if we switch before child completes (prevent memory leaks #2355):

      AssertionError: expected 2 to equal 1
      + expected - actual

      -2
      +1
      
      at spec/operators/switch-spec.ts:258:10
      at Context.modified (spec/helpers/testScheduler-ui.ts:206:13)
```

Reply:
I'm familiar with some of those tests. They are very low-level and fragile. It's likely that the change is unnecessary and that the low-level tests should be changed, instead.

I will have a close look at this later.

Thanks for the work that you've out into this PR.